File: frontend/src/shared/types/order.ts
import type { PricingSummary } from "./pricing";

export type OrderItem = {
  id: number;
  product_id: number | null;
  product_name: string;
  base_unit_price: number;
  quantity: number;
  unit_price: number;
  commercial_discount_amount: number;
  note: string | null;
};

export type CheckoutResponse = {
  order_id: number;
  status: string;
  payment_status: string;
  payment_reference: string | null;
  checkout_url: string | null;
};

export type CheckoutRequest = {
  store_id: number;
  address_id: number | null;
  delivery_mode: "delivery" | "pickup";
  payment_method: "cash" | "mercadopago";
};

export type Order = {
  id: number;
  store_id: number;
  store_name: string;
  store_slug: string;
  customer_name: string;
  delivery_mode: "delivery" | "pickup";
  payment_method: "cash" | "mercadopago";
  payment_status: string;
  payment_reference: string | null;
  status: string;
  address_label: string | null;
  address_full: string | null;
  store_latitude: number | null;
  store_longitude: number | null;
  address_latitude: number | null;
  address_longitude: number | null;
  subtotal: number;
  commercial_discount_total: number;
  financial_discount_total: number;
  delivery_fee: number;
  service_fee: number;
  delivery_fee_customer: number;
  rider_fee: number;
  total: number;
  delivery_status: string;
  delivery_provider: string;
  delivery_zone_id: number | null;
  assigned_rider_id: number | null;
  assigned_rider_name: string | null;
  assigned_rider_phone_masked: string | null;
  assigned_rider_vehicle_type: string | null;
  tracking_last_latitude: number | null;
  tracking_last_longitude: number | null;
  tracking_last_at: string | null;
  tracking_stale: boolean;
  eta_minutes: number | null;
  otp_required: boolean;
  merchant_ready_at: string | null;
  out_for_delivery_at: string | null;
  delivered_at: string | null;
  updated_at: string | null;
  created_at: string;
  items: OrderItem[];
  pricing: PricingSummary;
};

export type OrderTracking = {
  order_id: number;
  status: string;
  delivery_status: string;
  delivery_provider: string;
  tracking_enabled: boolean;
  assigned_rider_id: number | null;
  assigned_rider_name: string | null;
  assigned_rider_phone_masked: string | null;
  assigned_rider_vehicle_type: string | null;
  store_latitude: number | null;
  store_longitude: number | null;
  address_latitude: number | null;
  address_longitude: number | null;
  tracking_last_latitude: number | null;
  tracking_last_longitude: number | null;
  tracking_last_at: string | null;
  tracking_stale: boolean;
  eta_minutes: number | null;
  otp_required: boolean;
  otp_code: string | null;
};

export type PendingOrderReview = {
  order_id: number;
  store_name: string;
  delivered_at: string | null;
  rider_name: string | null;
  requires_rider_rating: boolean;
};

export type CreateOrderReviewPayload = {
  store_rating: number;
  rider_rating?: number | null;
  review_text?: string | null;
};

export type OrderStatusUpdate = {
  status:
    | "created"
    | "accepted"
    | "preparing"
    | "ready_for_dispatch"
    | "ready_for_pickup"
    | "out_for_delivery"
    | "delivered"
    | "cancelled"
    | "delivery_failed";
};

File: backend/app/schemas/order.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.pricing import PricingSummaryRead
from app.schemas.promotion import AppliedPromotionSummaryRead


class CheckoutRequest(BaseModel):
    store_id: int
    address_id: int | None = None
    delivery_mode: Literal["delivery", "pickup"]
    payment_method: Literal["cash", "mercadopago"]
    idempotency_key: str | None = Field(default=None, max_length=160)
    customer_latitude: float | None = None
    customer_longitude: float | None = None


class CheckoutResponse(BaseModel):
    order_id: int
    status: str
    payment_status: str
    payment_reference: str | None = None
    payment_transaction_id: int | None = None
    provider_preference_id: str | None = None
    checkout_url: str | None = None


class PaymentTransactionRead(BaseModel):
    id: int
    order_id: int
    provider: str
    external_reference: str
    preference_id: str | None = None
    payment_id: str | None = None
    status: str
    provider_status: str | None = None
    status_detail: str | None = None
    amount_total: float
    gross_amount: float
    marketplace_fee: float
    net_amount: float
    currency: str
    requested_marketplace_fee: float
    approved_marketplace_fee: float | None = None
    seller_expected_amount: float
    delivery_fee_amount: float
    service_fee_amount: float
    mp_user_id: str | None = None
    live_mode: bool | None = None
    checkout_url: str | None = None
    last_sync_at: datetime | None = None
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime | None = None


class OrderItemRead(BaseModel):
    id: int
    product_id: int | None = None
    product_name: str
    base_unit_price: float
    quantity: int
    unit_price: float
    commercial_discount_amount: float
    note: str | None = None


class OrderRead(BaseModel):
    id: int
    store_id: int
    store_name: str
    store_slug: str
    customer_name: str
    delivery_mode: str
    payment_method: str
    payment_status: str
    payment_reference: str | None = None
    status: str
    address_label: str | None = None
    address_full: str | None = None
    store_latitude: float | None = None
    store_longitude: float | None = None
    address_latitude: float | None = None
    address_longitude: float | None = None
    subtotal: float
    commercial_discount_total: float
    financial_discount_total: float
    delivery_fee: float
    service_fee: float
    delivery_fee_customer: float
    rider_fee: float
    total: float
    delivery_status: str
    delivery_provider: str
    delivery_zone_id: int | None = None
    assigned_rider_id: int | None = None
    assigned_rider_name: str | None = None
    assigned_rider_phone_masked: str | None = None
    assigned_rider_vehicle_type: str | None = None
    tracking_last_latitude: float | None = None
    tracking_last_longitude: float | None = None
    tracking_last_at: datetime | None = None
    tracking_stale: bool = False
    eta_minutes: int | None = None
    otp_required: bool = False
    merchant_ready_at: datetime | None = None
    out_for_delivery_at: datetime | None = None
    delivered_at: datetime | None = None
    updated_at: datetime | None = None
    created_at: datetime
    items: list[OrderItemRead]
    pricing: PricingSummaryRead
    applied_promotions: list[AppliedPromotionSummaryRead] = []


class OrderTrackingRead(BaseModel):
    order_id: int
    status: str
    delivery_status: str
    delivery_provider: str
    tracking_enabled: bool
    assigned_rider_id: int | None = None
    assigned_rider_name: str | None = None
    assigned_rider_phone_masked: str | None = None
    assigned_rider_vehicle_type: str | None = None
    store_latitude: float | None = None
    store_longitude: float | None = None
    address_latitude: float | None = None
    address_longitude: float | None = None
    tracking_last_latitude: float | None = None
    tracking_last_longitude: float | None = None
    tracking_last_at: datetime | None = None
    tracking_stale: bool = False
    eta_minutes: int | None = None
    otp_required: bool = False
    otp_code: str | None = None


class PendingOrderReviewRead(BaseModel):
    order_id: int
    store_name: str
    delivered_at: datetime | None = None
    rider_name: str | None = None
    requires_rider_rating: bool


class OrderReviewCreate(BaseModel):
    store_rating: int = Field(ge=1, le=5)
    rider_rating: int | None = Field(default=None, ge=1, le=5)
    review_text: str | None = None


class OrderStatusUpdate(BaseModel):
    status: Literal[
        "created",
        "accepted",
        "preparing",
        "ready_for_dispatch",
        "ready_for_pickup",
        "out_for_delivery",
        "delivered",
        "cancelled",
        "delivery_failed",
    ]


class MercadoPagoWebhookPayload(BaseModel):
    reference: str
    status: Literal["pending", "approved", "paid", "processing", "rejected", "refunded", "cancelled", "chargeback"]


class MercadoPagoPaymentSessionRead(BaseModel):
    session_token: str
    public_key: str
    order_id: int
    store_id: int
    store_name: str
    external_reference: str
    amount: float
    marketplace_fee: float
    net_amount: float
    currency: str = "ARS"
    status: str
    mode: str
    simulated: bool
    expires_at: datetime | None = None


class MercadoPagoCardPaymentPayerIdentification(BaseModel):
    type: str | None = None
    number: str | None = None


class MercadoPagoCardPaymentPayer(BaseModel):
    email: str
    identification: MercadoPagoCardPaymentPayerIdentification | None = None


class MercadoPagoCardPaymentRequest(BaseModel):
    session_token: str
    token: str | None = None
    issuer_id: str | int | None = None
    payment_method_id: str
    transaction_amount: float
    installments: int = Field(default=1, ge=1)
    payer: MercadoPagoCardPaymentPayer


class MercadoPagoCardPaymentResponse(BaseModel):
    order_id: int
    payment_id: str | None = None
    status: str
    provider_status: str | None = None
    status_detail: str | None = None
    external_reference: str
